### PR TITLE
Added Docker support to quickstart

### PR DESCRIPTION
**Pull request by bytesized**
@willkg Would you be a good reviewer for this?

### CODE REVIEW

File: quickstart.py [hunk 1]
```diff
@@ -46,18 +51,25 @@ def main():
     "config JSON file.")
   group.add_argument('--configJSON', metavar = "JSON", help = "Literal JSON to "
     "load configuration from rather than a configuration file.")
-  group.add_argument('--stop', action = "store_true", help = "If specified, "
+  parser.add_argument('--stop', action = "store_true", help = "If specified, "
```
Comment (bytesized) [inline quickstart.py:54]: Stop command now needs config to know whether it is stopping Docker.

File: quickstart.py [hunk 2]
```diff
@@ -46,18 +51,25 @@ def main():
...
     "this script will stop all servers rather than starting them. Note that "
     "this will not necessarily work if quickstart was not used to start the "
     "servers.")
+  parser.add_argument('--foreground', '-F', action = "store_true",
```
Comment (bytesized) [inline quickstart.py:58]: Docker needs quickstart to run in the foreground so that the container continues to run.

**Comment by willkg** [comment]
I can review it, but I probably need a list of things you want me to review and I still don't understand the usage scenarios (who will use it? how will they use it? in what context? etc).

**Comment by ddurst** [comment]
There is one, and potentially two usage scenarios for this:
1) by a developer for local symbol resolution (typically platform dev or anyone else looking into a crash report in-depth);
2) then, if this is more performant/stable than its predecessor, we would use this service to symbolicate crash stacks that we get from client crash pings (bug 1280469). We aren't entirely sure how the integration would be, so for now we want to verify it works well for scenario 1 and that it performs well under load.

**Comment by bytesized** [comment]
@willkg I was hoping you could review the whole PR, if possible.

File: quickstart.py [hunk 3]
```diff
@@ -186,7 +224,10 @@ def startServers(config):
     ]
     command.extend(configOptions)
     print "Starting SymServer with command: {}".format(command)
-    runDaemon(command)
+    if foregroundSymServer:
+      subprocess.call(command)
+    else:
+      runDaemon(command)
   return True
```
Comment (willkg) [inline quickstart.py:236]: I think it makes a lot of sense to reduce a lot of this code and use honcho instead.

https://honcho.readthedocs.io/en/latest/

It lets you define a Procfile and then runs the various things in that and handles all the process management. We use it on Socorro for development. I've used it for other web apps that have multiple processes, too.

File: quickstart.py [hunk 4]
```diff
@@ -261,14 +296,10 @@ def serversResponding(config):
       print "Timeout exceeded waiting for SymServer to start"
       return False
 
-    # Make sure we have a port value to contact SymServer on.
-    if 'configPath' in config:
-      SymServer.config.loadFile(config['configPath'])
-    elif 'configJSON' in config:
-      SymServer.config.loadJSON(config['configJSON'])
+    port = getSymServerConfig(config)['port']
```
Comment (willkg) [inline quickstart.py:304]: For configuration, I recommend using a library that can pull from INI files like you've got but also from environment variables. That makes this project more 12-factor friendly and will make it easier to spin up on servers.

We use python-decouple a lot: https://pypi.python.org/pypi/python-decouple

I wrote one over the summer that's similar to python-decouple but lacks some of the things I found irritating. Having said that, I'm not sure you have the same needs I had and mine is alpha-ish still.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,23 @@
+FROM python:2.7.12
```
Comment (willkg) [inline Dockerfile:1]: Python 2.7 has a very short shelflife now. If you use Python 2.7, you should expect to do a rewrite in the next couple of years.

I've been doing all new things with Python 3.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+# Setup environment/dependencies
+RUN update-alternatives --install /bin/sh sh /bin/bash 10
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+                    build-essential \
+                    memcached && \
+    rm -rf /var/lib/apt/lists/*
+RUN groupadd --gid 1001 app && useradd -g app --uid 1001 --shell /usr/sbin/nologin app
+COPY . /app
+RUN pip install -r /app/requirements.txt
```
Comment (willkg) [inline Dockerfile]: To make this more prod-secure, you'll want to use hashes in the requirements file and use a version of pip that supports hashes.

An example of a requirements file with hashes in Antenna:

https://github.com/mozilla/antenna/blob/master/requirements.txt

Any example of upgrading pip to a version that supports hash checking before installing requirements:

https://github.com/mozilla/antenna/blob/master/Dockerfile#L11

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+RUN chown -R app:app /app
```
Comment (willkg) [inline Dockerfile:20]: Why chown everything here?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+# Setup environment/dependencies
+RUN update-alternatives --install /bin/sh sh /bin/bash 10
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+                    build-essential \
+                    memcached && \
+    rm -rf /var/lib/apt/lists/*
+RUN groupadd --gid 1001 app && useradd -g app --uid 1001 --shell /usr/sbin/nologin app
+COPY . /app
```
Comment (willkg) [inline Dockerfile:17]: This is a good idea. Typically, I mount `.` as a volume so then I can access it and any changes I make to it in the docker container. For the use cases you're targeting, if the code changes I think you want someone to rebuild their docker container and thus wipe out any state along with it. Thus mounting `.` as a volume like I typically do, just muddies things.

File: README.md [hunk 9]
```diff
@@ -8,20 +8,42 @@ server-side symbol files (.SYM files).
...
-Quick Start Guide
------------------
+Quick Start without Docker
+--------------------------
 
 1. Install [Python 2.7](https://www.python.org/downloads/) if it is not already
    installed
-2. Install dependencies:
-   `python -m pip install tornado futures python-memcached psutil`
+2. Install dependencies: `python -m pip install -r requirements.txt`
```
Comment (willkg) [inline README.md]: If python is in the path and pip is installed, then pip will also be in the path, so you can just do:

```
pip install -r requirements.txt
```

Also, pip doesn't come with Python, so if someone is installing Python fresh, they're going to need to install pip, too.

I think this covers it: https://pip.pypa.io/en/stable/installing/

File: README.md [hunk 10]
```diff
@@ -8,20 +8,42 @@ server-side symbol files (.SYM files).
...
+Quick Start with Docker
+-----------------------
+
+1. Install [Docker](https://docs.docker.com/engine/installation)
+2. It may be necessary that your user be in the Docker group. This can be
+   accomplished in Linux with `sudo usermod -aG docker $(whoami)`. You must log
+   out for changes to take effect.
+3. Install [Python 2.7](https://www.python.org/downloads/) if it is not already
+   installed
+4. Install dependencies: `python -m pip install -r requirements.txt`
```
Comment (willkg) [inline README.md]: I commented earlier on a similar line--the same thing applies here.

File: README.md [hunk 11]
```diff
@@ -8,20 +8,42 @@ server-side symbol files (.SYM files).
...
+   [Configuration File](#configuration-file). To use Docker, it is important
+   that the `quickstart.Docker.enable = true`.
+6. Run `python quickstart.py -c [configuration file]`
+
+Quickstart attempts to do some caching in order to recreate the image and
+container only when necessary. The image will be built only once. The container
+will be recreated any time that the passed configuration changes. You can
+force a full rebuild with `python quickstart.py -c [configuration file] -R`.
+The docker container can be stopped with
+`python quickstart.py -c [configuration file] --stop`.
```
Comment (willkg) [inline README.md:56]: The way you have the Dockerfile set up, if they ever update the code, they're going to want to do a rebuild of the container. Probably helpful to mention that here.

File: README.md (hunk 9 above)
Comment (willkg) [reply README.md]: Also, this line has people installing requirements into their system Python. You probably want them to create virtual environments so that the library dependencies of this project don't interfere with those of other projects.

There's more information on virtual environments here:

http://docs.python-guide.org/en/latest/dev/virtualenvs/

File: README.md [hunk 12]
```diff
@@ -188,6 +210,31 @@ All configuration values should be specified as strings unless otherwise noted.
...
 - `"quickstart"` Configuration options related to the quickstart script
+    - `"Docker"` Configuration options related to Docker
+        - `"enable"` Must be a boolean type. If set to `true`, quickstart
+          will start the servers within a Docker container
+        - `"restart"` Must be a boolean type. If set to `false`, quickstart
+          will not restart the Docker container if it is already started.
+          This is, however, subject to some conditions. If a rebuild is needed
+          (or requested with `-R`), or if the configuration options have
+          changed, the Docker container will be restarted regardless of this
+          option.
```
Comment (willkg) [inline README.md]: What's the use case for having an explicit restart and not just doing stop and start again by hand?

File: quickstart.py [hunk 13]
```diff
@@ -309,6 +340,196 @@ def sendGetRequest(port):
...
+  if not config["Docker"]["restart"] and not forceStop:
+    print "Not stopping Docker because it is not configured for restart"
+    return True
+
+  dockerApi = getDockerApi(config)
+  try:
+    dockerApi.stop(container = DOCKER_CONTAINER_NAME)
+  except:
+    print "Failed to stop Docker"
+    return False
```
Comment (willkg) [inline quickstart.py:358]: This is a bare try/except and it's not great. The print statement doesn't tell you anything about the exception. For example, this could be dying because you spelled `DOCKER` wrong in the variable value.

Given that you don't know what you want to catch here, it's better for your sanity to do something like this:

```
try:
    dockerApi.stop(container=DOCKER_CONTAINER_NAME)
except Exception as exc:
    print 'Failed to stop Docker:', exc
    return False
```

Comment (willkg) [reply quickstart.py:358]: This is true of all the other try/except that don't catch any exceptions, too.

File: quickstart.py [hunk 14]
```diff
@@ -309,6 +340,196 @@ def sendGetRequest(port):
...
+  oldDockerConfig = getOldDockerConfigString()
+  imageInfo = dockerImageInfo(config)
+  rebuild = forceRebuild or not imageInfo or oldDockerConfig == None
+  if rebuild:
+    print "Building Docker image"
+    try:
+      dockerApi.remove_image(image = DOCKER_IMAGE_NAME, force = True)
+    except:
+      # Maybe the image just doesn't exist yet
+      pass
```
Comment (willkg) [inline quickstart.py:391]: This try/except form has a name called "the diaper pattern":

http://mike.pirnat.com/2009/05/09/the-diaper-pattern-stinks/

I mention it because it's kind of a funny anti-pattern name.

File: quickstart.py [hunk 15]
```diff
@@ -309,6 +340,196 @@ def sendGetRequest(port):
...
+
+    for output in dockerApi.build(path = SRC_DIR, tag = DOCKER_IMAGE_NAME):
+      pass
```
Comment (willkg) [inline quickstart.py]: This prevents you from seeing any issues in the build output. That's ... probably not wildly helpful if things go awry. I can see why you might not want this spitting tons of stuff out all the time, though.

Maybe this is a good reason to add a logging level configuration variable?

File: quickstart.py [hunk 16]
```diff
@@ -309,6 +340,196 @@ def sendGetRequest(port):
...
+  """ Returns the configuration used to build the last docker container as a
+  string. Returns |None| if there is no old configuration
+  """
+  data = None
+  try:
+    with open(DOCKER_CACHE_FILE, 'r') as fp:
+      data = fp.read()
+  except:
+    return None
+  return data
```
Comment (willkg) [inline quickstart.py]: You could write this more simply as:

```
try:
    with open(DOCKER_CACHE_FILE, 'r') as fp:
        return fp.read()
except (OSError, IOError):
    pass
return None
```

Comment (willkg) [reply quickstart.py]: This is true of the various other similarly structured blocks.

File: requirements.txt [hunk 17]
```diff
@@ -1,4 +1,5 @@
 tornado
 futures
 python-memcached
-psutil
\ No newline at end of file
+psutil
+docker-py
```
Comment (willkg) [inline requirements.txt]: If you want to go to production with this, you'll need to pin these. Additionally, you should add hashes for them.

File: snappy/DiskCache_Config.py [hunk 18]
```diff
@@ -1,5 +1,6 @@
 import os
 import json
+from configUpdate import configUpdate
```
Comment (willkg) [inline snappy/DiskCache_Config.py:3]: If you're going to do a relative import, then this should be:

```
from .configUpdate import configUpdate
```

However, I don't think these files will ever exist outside of snappy, so I'd do this instead:

```
from snappy.configUpdate import ConfigUpdate
```

Comment (willkg) [reply snappy/DiskCache_Config.py:3]: This is true of the other `configUpdate` imports elsewhere in this PR.

File: snappy/configUpdate.py [hunk 19]
```diff
@@ -0,0 +1,9 @@
+# Updates a configuration with new values
+def configUpdate(original, update):
+  for key in update:
+    if (key in original and
+        isinstance(original[key], dict) and
+        isinstance(update[key], dict)):
+      configUpdate(original[key], update[key])
+    else:
+      original[key] = update[key]
```
Comment (willkg) [inline snappy/configUpdate.py:21]: This could use some tests.

Comment (willkg) [reply snappy/configUpdate.py:21]: One thing this does is if the two configs have different shapes, the result is whatever the update is. Is that what you want? Would it be better for the `configUpdate` to raise an error when the two dicts have different shapes figuring that indicates something has gone awry?

**Review by willkg — CHANGES_REQUESTED** [review]
This is an interesting approach. There's a lot of stuff here that's different than how i would do it, but I'm not sure that's an error and i don't think there are important consequences for doing it differently. It just means it's harder for me to get a mental model working.

There are some python-y things I called out. I added some comments about things we should be thinking about before putting it in production.

I tested it out locally very briefly. Seems like it works fine, but I'm on Linux and that's sort of a magical unicorn-infested place for Docker things. It'd be good to have someone test this out on OSX as well as on Windows.

I'm impressed @bytesized banged a docker-compose equivalent out so quickly.

That's everything I can think of. I hope it's helpful!

File: quickstart.py (hunk 4 above)
Comment (willkg) [reply quickstart.py:304]: Oops--I forgot you're using JSON. I'd switch to INI files or env files something else that supports comments.

**Comment by willkg** [comment]
Also, before you go to prod, you really want to set up a CI system (Circle, Travis, Jenkins, etc) to run tests on every PR. That's worth setting up early in a project, too, since it makes it easier to accept contributions from other people.

File: quickstart.py (hunk 3 above)
Comment (bytesized) [reply quickstart.py:236]: I'm not sure that I totally understand honcho, but it doesn't seem like it addresses my needs. I need to have the option of running in the foreground OR background. Running with Docker requires that the process be in the foreground while the container is to remain open. Running without Docker, you probably do not want Snappy to run in the foreground. To me it looks like honcho does not allow for this flexibility.

File: quickstart.py (hunk 4 above)
Comment (bytesized) [reply quickstart.py:304]: Perhaps loading configuration from environment variables is desirable, but I think that should be filed as an issue rather than part of this pull request. It is not really related to Docker.

File: Dockerfile (hunk 5 above)
Comment (bytesized) [reply Dockerfile:1]: Right now the app is written for Python 2.7. I think that rewriting in Python 3 may need to be done, but it is outside the scope of this pull request.

File: README.md (hunk 12 above)
Comment (bytesized) [reply README.md]: I'm not sure what you mean by this. There is no explicit restart option. There is an explicit rebuild option.

**Comment by bytesized** [comment]
Setting up Travis CI is next on my todo list.

File: quickstart.py (hunk 4 above)
Comment (bytesized) [reply quickstart.py:304]: I also don't think that it is worth moving to INI just for comments. Especially given that JSON provides types and nesting.

File: README.md (hunk 9 above)
Comment (bytesized) [reply README.md]: Freshly installed Python does come with pip: [https://pip.pypa.io/en/stable/installing/](https://pip.pypa.io/en/stable/installing/)
But it is possible that they user still might not have it installed, so I'll add it.

File: quickstart.py (hunk 4 above)
Comment (willkg) [reply quickstart.py:304]: I was told to review everything with a secondary use case of pushing this to production. Antenna runs in production in a Docker container, thus I thought this was relevant.

File: quickstart.py (hunk 3 above)
Comment (willkg) [reply quickstart.py:236]: You just put honcho in the background:

```
$ honcho &
```

File: snappy/DiskCache_Config.py (hunk 18 above)
Comment (bytesized) [reply snappy/DiskCache_Config.py:3]: This is currently how all internal imports work. It turns out that changing it would require some significant changes. I think that this needs to be filed as a separate issue.

File: README.md (hunk 9 above)
Comment (willkg) [reply README.md]: Oh nice--I didn't know that.

File: README.md (hunk 12 above)
Comment (willkg) [reply README.md]: The text here says:

```
 +        - `"restart"` Must be a boolean type. If set to `false`, quickstart 
```

Isn't that an explicit restart option?

Comment (bytesized) [reply README.md]: Oh, I misunderstood. I originally put it in for consistency with the other quickstart sub-options. I think you are right though. I will remove it.

File: Dockerfile (hunk 7 above)
Comment (bytesized) [reply Dockerfile:20]: Because logger and DiskCache need write permissions in the directory.

**Comment by bytesized** [comment]
@willkg Ok, what do you think of my revisions?

File: requirements.txt [hunk 20]
```diff
@@ -1,5 +1,29 @@
...
+six==1.10.0 \
+    --hash=sha256:0ff78c403d9bccf5a425a6d31a12aa6b47f1c21ca4dc2573a7e2f32a97335eb1
+requests==2.10.0 \
+    --hash=sha256:09bc1b5f3a56cd8c48d433213a8cba51a67d12936568f73b5f1793fcb0c0979e
+ipaddress==1.0.17 \
+    --hash=sha256:a42764805223838ec9d04eee7fe699083bf5321001792b6076322784c5b994a5
+websocket-client==0.37.0 \
+    --hash=sha256:678b246d816b94018af5297e72915160e2feb042e0cde1a9397f502ac3a52f41
+docker-pycreds==0.2.1 \
+    --hash=sha256:58d2688f92de5d6f1a6ac4fe25da461232f0e0a4c1212b93b256b046b2d714a9
```
Comment (willkg) [inline requirements.txt:29]: @peterbe wrote hashin which makes these a lot easier to maintain:

https://pypi.python.org/pypi/hashin

Figured I'd mention it to ease future maintenance burden.

File: snappy/configUpdate.py [hunk 21]
```diff
@@ -1,9 +1,21 @@
...
+        raise TypeError("Attempt to update key {} with a non list type".format(key))
+      original[key] = update[key]
+    elif isinstance(original[key], int):
+      original[key] = int(update[key])
+    elif isinstance(original[key], float):
+      original[key] = float(update[key])
+    elif isinstance(original[key], basestring):
+      if not isinstance(update[key], basestring):
+        raise TypeError("Attempt to update key {} with a non string type".format(key))
       original[key] = update[key]
```
Comment (willkg) [inline snappy/configUpdate.py:21]: This should be an easy function to add tests for--especially the edge cases. Definitely worth doing.

**Comment by willkg** [comment]
This looks good enough.

One thing I noticed is that this adds a lot of non-trivial code that has no unit tests and requires manual testing. That's definitely something to think about because it's creating a non-trivial maintenance burden. It'd be worth writing out the tests you're doing by hand so that someone who makes changes to the code that's not covered by the unit tests can go through and methodically verify no new issues were created. You could store that in the project wiki or in a `.rst/.md` file in the repo.

**Review by willkg — APPROVED** [review]
